packages/sdk-codegen/src/specLinter.ts

```diff
@@ -0,0 +1,203 @@
+  const diff: TypeDelta = {}
+  Object.entries(lType.properties).forEach(([key, lhsVal]) => {
+    const rhsVal = rType.properties[key]
+    if (rhsVal) {
+      const lhs = lhsVal.summary()
+      const rhs = rhsVal.summary()
+      if (lhs !== rhs) {
+        diff[key] = {
+          lhs,
+          rhs,
+        }
+      }
+    }
+  })
+  if (Object.keys(diff).length === 0) return
+  return diff
+}
+
+/**
+ * Compares the params of two methods and returns the signature of both methods
+ * if non matching.
+ * @param lMethod
+ * @param rMethod
+ */
+export const compareParams = (lMethod: IMethod, rMethod: IMethod) => {
+  const lSignature = lMethod.signature()
+  const rSignature = rMethod.signature()
+
+  if (lSignature === rSignature) return
+
+  return {
+    lhs: lSignature,
+    rhs: rSignature,
+  }
+}
+
+export const csvHeaderRow =
+  'method name,Op + URI,left status,right status,typeDiff,paramDiff'
+
+export const csvDiffRow = (diff: DiffRow) => `
+${diff.name},${diff.id},${diff.lStatus},${diff.rStatus},${diff.typeDiff},${diff.paramsDiff}`
+
+export const mdHeaderRow = `
+| method name  | Op + URI | 3.1 | 4.0 | typeDiff
+| ------------ | -------- | --- | --- | ------------`
+
+export const mdDiffRow = (diff: DiffRow) => `
+  | ${diff.name} | ${diff.id} | ${diff.lStatus} | ${diff.rStatus} | ${diff.typeDiff} | ${diff.paramsDiff} |`
+
+const diffToString = (diff: any) =>
+  diff ? `"${JSON.stringify(diff, null, 2).replace(/"/g, "'")}."` : ''
+
+export interface ComputeDiffInputs {
+  lMethod?: IMethod
+  rMethod?: IMethod
+}
+
+interface DiffRow {
+  /** Method name */
+  name: string
+  /** Method operation and path */
+  id: string
+  /** Method status in both specs */
+  lStatus: string
+  rStatus: string
+  /** Method type diff if any */
+  typeDiff: string
+  /** Method params diff if any */
+  paramsDiff: string
+}
+
+/**
+ * Computes the diff of two methods
+ * @param lMethod
+ * @param rMethod
+ */
+const computeDiff = ({ lMethod, rMethod }: ComputeDiffInputs) => {
+  if (!lMethod && !rMethod) throw new Error('At least one method is required.')
+
+  let result: DiffRow
+  if (lMethod) {
+    result = {
+      name: lMethod.name,
+      id: lMethod.id,
+      lStatus: lMethod.status,
+      rStatus: rMethod ? rMethod.status : '',
+      typeDiff: rMethod
+        ? diffToString(compareTypes(lMethod.type, rMethod.type))
+        : '',
+      paramsDiff: rMethod ? diffToString(compareParams(lMethod, rMethod)) : '',
+    }
+  } else if (!lMethod && rMethod) {
+    result = {
+      name: rMethod.name,
+      id: rMethod.id,
+      lStatus: '',
+      rStatus: rMethod.status,
+      typeDiff: '',
+      paramsDiff: '',
+    }
+  }
+
+  return result!
+}
+
+/**
+ * Given two specs, compares each method's params and response type, and returns
+ * a string representing representing a csv containing the difference if any
+ * @param lSpec
+ * @param rSpec
+ */
+
+export type preDiffCb = ({ lMethod, rMethod }: ComputeDiffInputs) => boolean
```

Comment:
Not sure these params match to the type signature now, and the blank line should be removed

Reply:
My mistake. That's an extra comment that should be removed.